### ISSUE TITLE
Make number of worker threads setable

### ISSUE THREAD

**Issue by ratemyduck** [matiasdelellis/facerecognition#237]
I have a server with 96GB memory and 12 cores (6 per processor - real cores, not hyperthreading cores). Naturally I want to have the load distributed among those cores. Thus it would be nice to have a "number of worker threads" parameter for the face:background_job. 

If there were a "only scan given directory" I guess I could write some small script that spawns and waits for a number of processes. But it would be much prettier to add it in the code of the plugin.

**Comment by stalker314314** [comment]
My thoughts... we cannot easily add support for threads into `face:background_job` as those jobs could be run in either cron, but also from web (depending how user configured background jobs). We could put support for `pthreads` or `parallel` under some conditions (it is is running from console, it makes sense to use more than 1 thread, there is config for this turned on), but it would require bit of refactoring. Let's assume that we don't want to parallelize everything, but just CNN model of face detection. In that case, changes would be localized to https://github.com/matiasdelellis/facerecognition/blob/master/lib/BackgroundJob/Tasks/ImageProcessingTask.php. We would also have some kind of abstract function in model that says `maxDOP` (degree of parallelism) that can be used on a given model. Doable, but there are lot of touchy parts to take care.

Another option that we can do, and I think it is more in spirit of PHP is to shard execution in cron jobs, so each user can control what and how to shard it. For example, today you can do sharding per user (enumerate all users in system, and spawn new `face:background_job` for each of them). Well, not exactly, as we have global lock, but you can remove it by commenting `LockTask` from [here](https://github.com/matiasdelellis/facerecognition/blob/master/lib/BackgroundJob/BackgroundService.php#L95). This is one way to do it. Other is that we add some more granular support for sharing. Maybe we can add some parameter to `face:background_job` that says "do only those images whose hash of absolute path ends with this string". So, if you give it "5df", only hashes of that ends with "5df" will be processed. We can even wrap this to be easier to user, by letting them specify two parameters - `count_shards` and `current_shard`. So, it can proceed by spawning 3 processes:
`face:background_job --current_shard 1 --count_shards 3`
`face:background_job --current_shard 2 --count_shards 3`
`face:background_job --current_shard 3 --count_shards 3`

This looks like better solution than introducing pthreads and parallel library to codebase. My thought:)

**Comment by ratemyduck** [comment]
I just looked very briefly at the code so my feedback would probably be of limited use.

I found the maps background task (it reads the geo metadata from images) to be the best strategy out there - instead of doing the work it schedules tasks with the regular system cron. However - that one does not parallelize either for some strange reason.

I got surprise by this: 
> .... as those jobs could be run in either cron, but also from web (depending how user configured background jobs)

I did not know it automatically hooked into the normal cron. You mean everything would eventually have been scanned even if I wound't have run face:background_job explicitly? 
But this do not matter for the issue at hand - I'm just curious. 

> Let's assume that we don't want to parallelize everything, but just CNN model of face detection. 
I concur. The clustering didn't take very long time.

> maxDOP 
A comment on naming. DOP is a term that isn't used very often. I would use "threads" or "workers".

> and I think it is more in spirit of PHP
What is the spirit of PHP in this context? To fire of more php processes rather than multithread inside one? 

As long as I as a user can fire of one command, I'm a happy camper. The proposed solution forces me to write a small script firing of a number of these - the implementation kinda "leaks" out onto the user which doesn't feel that good. 

Like I said before - it's a pity the normal NC cron does not parallellize - then it would just have been a matter of scheduling all the images to be processed. All the heavy lifting would have been taken off the plugin developers.  

And lastly, a confession - I got a bit of problem with the gui loading very slowly after the installation of this plugin, so I uninstalled it again. (It wasn't the problem with pdlib causing havoc in the fpm process - I only added it in the CLI config for php.) Meaning - I posted this ticket because I think it might be a good thing for your users - of which I'm currently not one.

**Comment by matiasdelellis** [comment]
Hi @ratemyduck 

> I posted this ticket because I think it might be a good thing for your users - of which I'm currently not one.

Oh .. A shame .. :sweat_smile: 
However it is something that is always considered, but still not within the priorities .. :wink: 

>  And lastly, a confession - I got a bit of problem with the gui loading very slowly after the installation of this plugin, so I uninstalled it again. (It wasn't the problem with pdlib causing havoc in the fpm process - I only added it in the CLI config for php.) Meaning - 

Please, if you can help us with this, which seems much more worrisome to me. :confused: 

I understand that it can slow down the server if the microprocessor is busy during the analysis (..and multiple threads would make this worse ..), however I never found differences regarding the functioning of the server only by the installation of the application or its dependencies. .

Can you describe a little more how it affected you, and how you corroborated it?

Thanks!. :smile:

**Comment by rhatguy** [comment]
I wanted to bring this back up.  I have a dual xeon with 16 cores per socket for a total of 32 real cores.  I just copied over ~1700 pictures into a test user and ran the initial scan against just that user.  Obviously, it is using a single thread (one core) and is estimating its going to take 4 days to process these 1700 images which are really just a test for me.  If I move this into my full photo collection of over 40,000 pictures I estimate it will take over 100 days.

It seems that the ImageProcessingTask should be relatively parallelize-able given that its the same process on each image.  I understand that clustering is dependent on relationships between images.  Is there some reason I couldn't just fire off 32 separate processes processing 32 separate images and let each one update the database with its findings for that image, then pull another image off of the queue.

The current single threaded nature of the image processing task makes this tool very difficult to get started with.

**Comment by cliffalbert** [comment]
In my setup with 173000 pictures, the scanning it self took around 4 days. However everytime a few pictures are added the clustering of faces takes around 8 hours before new scanning is started. 

In my use case it would be beneficial if clustering could be multi-threaded

**Comment by rhatguy** [comment]
@cliffalbert how are you scanning 173k pictures in 4 days?  Are you referring to step 9/10 which is the ImageProcessingTask?  Are you using a GPU, and are you using model 4?  My server is spending significant amounts of time on each picture in a serial fashion on this step. I haven't even gotten to the point to start clustering yet (don't yet have 1000 faces) and its estimating 4 days for this current step.

**Comment by matiasdelellis** [comment]
Hi both,

> The current single threaded nature of the image processing task makes this tool very difficult to get started with.

@rhatguy 
First of all I will answer you this. :wink: 

I understand that it can be frustrating.. :disappointed: 
However, I understand that such a number of photos is after many years. It may take 100 days (And still agreeing that it is slow.) to analyze them, but once it is finished, you will not add so many photos. Therefore, you will upload 10 photos in an event, and it will take 10 minutes to see the results. Which would be acceptable. Not even google photos give you instant results. 

> It seems that the ImageProcessingTask should be relatively parallelize-able given that its the same process on each image

In principle, it may seem easy, however, php is not as multi-threaded as you might think. Just in 7.2 they added something more or less official for it.
* https://www.php.net/manual/es/intro.parallel.php

Maybe with this we can implement it, but still don't have time to investigate. We will have to wait. :wink: 

Please, note that pthreads It's not even maintained! :disappointed: 
* https://github.com/krakjoe/pthreads/issues/929

>  Is there some reason I couldn't just fire off 32 separate processes processing 32 separate images

The main reason in any case is ram memory. Today anyone has microprocessors with several cores, but ram memory is still expensive. In my case, I have 8 cores, but only 8 Gb of ram. :disappointed: :cry: :sob: 
However, I still recognize that this feature will be implemented.. :wink: 

@cliffalbert 
> In my setup with 173000 pictures, the scanning it self took around 4 days.

Wow, awesome. I guess model 1, with cuda enabled.

> However everytime a few pictures are added the clustering of faces takes around 8 hours before new scanning is started.

This is something that worries me much more. :disappointed: 

What version of pdlib are you using? v1.0.2 includes changes that make it much faster.
* https://github.com/matiasdelellis/facerecognition/issues/105

**Comment by cliffalbert** [comment]
> @cliffalbert how are you scanning 173k pictures in 4 days? Are you referring to step 9/10 which is the ImageProcessingTask? Are you using a GPU, and are you using model 4? My server is spending significant amounts of time on each picture in a serial fashion on this step. I haven't even gotten to the point to start clustering yet (don't yet have 1000 faces) and its estimating 4 days for this current step.

Step 9/10 is taking not quite that long. It's mainly step 6/10 that takes a lot of time the 2nd run. Altough I had a lot of interruptions in 9/10 due to NextCloud's image type recognition (which causes crashes sometimes) 9/10 itself does run quite fast on my machine.

> @matiasdelellis 

No Cuda, model 1. 

It seems still pdlib 1.0.1, i'll go and upgrade that one to 1.0.2

**Comment by cliffalbert** [comment]
Great! That reduced the run time to about 2 hours :)

**Comment by MaartenMees** [comment]
Hi, I have also threadripper and 96gb ram. And just started the analysis of 50k of photos. It's gonna take a month...
Any progress in this topic? :)
Thanks!

**Comment by rhatguy** [comment]
Bringing this back up.  With the integration with memories I've circled back to face recognition and even in my initial testing face recognition's "correctness" in clustering is the best of any tools I've tried.  Running on unraid on a dual Xeon with 28 real cores now and again struggling with the length of time it takes to scan pictures.  Currently using face recognition 0.9.10 with method 4 and when I do a scanning run I notice the most time is spent on step 8/8 (ImageProcessingTask) where face recognition is selecting pictures seemingly at random from the list of photos to be scanned.  Looking at top its mostly using 1 core.

Is there some reason we couldn't kick off multiple ImageProcessingTask processes (maybe hundreds) that would each work on a single picture (even if it was a one time command that ran manually outside of the normal cron/php structure)?  As has been mentioned before, the need to bulk scan thousands of images is an initial, one time problem and not an ongoing one.  Is there some interdependence on step 8 where the output of one pictures scan might affect the way another picture is scanned?  I understand the dependence during clustering, but don't understand the why the individual pictures need to be scanned serially and just curious if its an environmental limitation or a limitation in the face scanning process itself.

**Comment by matiasdelellis** [comment]
Wow, After 4 years we have something about it.

![imagen](https://github.com/matiasdelellis/facerecognition/assets/733715/e6fe68bb-1800-45f8-9532-902cb9916039)

Dy defaultt, the background_job command continues to work as usual, and is recommended, but in the latest commits 3 special modes are added.

`$ occ face:background_job -u user --sync-mode`: It must be run only in one thread, and is basically only dedicated to finding new files.

`$ occ face:background_job -u user --analyze-mode`:  This mode only executes the hard processing of the images, and since the last commit it can be executed in parallel several times. 😬 

`$ occ face:background_job -u user --cluster-mode`: Finally this mode is responsible for obtaining people from the faces. AKA clustering. It is not thread safe either..

**NOTE:**
Of course one must know how to manage resources to know how many threads can execute these command. On the other hand, I must clarify that the external model does not yet allow parallel execution. We must work on that.
